pkg/api/validation/events.go

```diff
@@ -25,20 +31,55 @@ import (
 // ValidateEvent makes sure that the event makes sense.
 func ValidateEvent(event *api.Event) field.ErrorList {
 	allErrs := field.ErrorList{}
-	// There is no namespace required for node.
+	// There is no namespace required for root-scoped kind, for example, node.
 	// However, older client code accidentally sets event.Namespace
 	// to api.NamespaceDefault, so we accept that too, but "" is preferred.
-	if event.InvolvedObject.Kind == "Node" &&
-		event.Namespace != api.NamespaceDefault &&
-		event.Namespace != "" {
-		allErrs = append(allErrs, field.Invalid(field.NewPath("involvedObject", "namespace"), event.InvolvedObject.Namespace, "not allowed for node"))
-	}
-	if event.InvolvedObject.Kind != "Node" &&
-		event.Namespace != event.InvolvedObject.Namespace {
-		allErrs = append(allErrs, field.Invalid(field.NewPath("involvedObject", "namespace"), event.InvolvedObject.Namespace, "does not match involvedObject"))
+	// Todo: Events may reference 3rd party object, and we can't check whether the object is namespaced.
+	// Suppose them are namespaced. Do check if we can get the piece of information.
+	// This should apply to all groups served by this apiserver.
+	group := apiutil.GetGroup(event.InvolvedObject.APIVersion)
+	version := apiutil.GetVersion(event.InvolvedObject.APIVersion)
+	if (group == "" || group == "extensions") && (version != "extensions" && version != "") {
+		namespacedKindFlag, err := isNamespacedKind(event.InvolvedObject.Kind, event.InvolvedObject.APIVersion)
+		if err != nil {
+			allErrs = append(allErrs, field.Invalid(field.NewPath("involvedObject", "kind"), event.InvolvedObject.Kind, fmt.Sprintf("couldn't check whether namespace is allowed: %s", err)))
+		} else {
+			if !namespacedKindFlag &&
+				event.Namespace != api.NamespaceDefault &&
```

Comment:
I think we should do the following:
* Do not require that an `Event.InvolvedObject.Namespace` must match the `Event.Namespace`
 * I do not understand the rationale in requiring that equality check generally.  
 * Why can't I create an `Event` in my namespace that references the `kubernetes` service in the default namespace?
 * Why can't I create an `Event.Namespace=Foo` whose `Event.InvolvedObject.Kind=Namespace` and `Event.InvolvedObject.Name=Bar`?  The current model basically makes it that I can't create an event about a namespace, which seems silly.
* we should fail validation if `Event.Namespace` is not specified
* we should not require `Event.InvolvedObject.Namespace` if the object is not-namespaced
* the `kubelet` should spawn node events in the namespace that is provided via configuration, maybe we can re-purpose `MasterServiceNamespace` for that even though it has a bad name.  we could then deprecate the endpoint to create non-namespaced events.

Every time I see these issues, I wish we did not put `Namespace` on `ObjectMeta`, and that we had made a `NamespaceMeta` that was added to namespaced-kinds.